cpp/docs/DEVELOPER_GUIDE.md

```diff
@@ -0,0 +1,977 @@
+
+### Scalar
+
+A type representing a single element of a data type.
+
+### Table
+
+A table is a collection of columns. It is also known as a dataframe.
+
+### View
+
+A view is a non-owning object that provides zero-copy access (possibly with slicing or offsets) data 
+owned by another object. Examples are column views and table views. 
+
+# Directory Structure and File Naming
+
+External/public libcudf APIs are grouped based on functionality into an appropriately titled 
+header file  in `cudf/cpp/include/cudf/`. For example, `cudf/cpp/include/cudf/copying.hpp` 
+contains the APIs for functions related to copying from one column to another. Note the  `.hpp` 
+file extension used to indicate a C++ header file.
+
+Header files should use the `#pragma once` include guard. 
+
+The naming of external API headers should be consistent with the name of the folder that contains 
+the source files that implement the API. For example, the implementation of the APIs found in
+`cudf/cpp/include/cudf/copying.hpp` are located in `cudf/src/copying`. Likewise, the unit tests for 
+the APIs reside in `cudf/tests/copying/`.
+
+Internal API headers containing `detail` namespace definitions that are used across translation 
+units inside libcudf should be placed in `include/cudf/detail`.
+
+## File extensions
+
+- `.hpp` : C++ header files
+- `.cpp` : C++ source files
+- `.cu`  : CUDA C++ source files
+- `.cuh` : Headers containing CUDA device code
+
+Only use `.cu` and `.cuh` if necessary. A good indicator is the inclusion of `__device__` and other
+symbols that are only recognized by `nvcc`. Another indicator is Thrust algorithm APIs with a device
+execution policy (always `rmm::exec_policy` in libcudf).
+
+## Code and Documentation Style and Formatting
+
+libcudf code uses [snake_case](https://en.wikipedia.org/wiki/Snake_case) for all names except in a 
+two cases: template parameters and unit tests and test case names may use Pascal case, aka 
+[UpperCamelCase](https://en.wikipedia.org/wiki/Camel_case). We do not use [Hungarian notation](https://en.wikipedia.org/wiki/Hungarian_notation), except sometimes when naming device data variables and their corresponding
+host copies. Private member variables are typically prefixed with an underscore.
+
+```c++
+template <typename IteratorType>
+void algorithm_function(int x, rmm::cuda_stream_view s, rmm::device_memory_resource* mr) 
+{
+  ...
+}
+
+class utility_class 
+{
+  ...
+ private:
+  int _rating{};
+  std::unique_ptr<cudf::column> _column{};
+}
+
+TYPED_TEST_CASE(RepeatTypedTestFixture, cudf::test::FixedWidthTypes);
+
+TYPED_TEST(RepeatTypedTestFixture, RepeatScalarCount)
+{
+  ...
+}
+```
+
+C++ formatting is enforced using `clang-format`. You should configure `clang-format` on your 
+machine to use the `cudf/cpp/.clang-format` configuration file, and run `clang-format` on all 
+changed code before committing it. The easiest way to do this is to configure your editor to 
+"format on save".
+
```

Comment:
How about some guidelines for `#include` statements? 
From @harrism slack message in November:

1. Group your includes by library (e.g. cuDF, RMM, Thrust, STL)
2. Separate groups by a blank line.
3. Order the groups from "nearest" to "farthest"
4. Use `<>` instead of `""` unless the header is in the same directory as the source file.
5. Let clang-format sort the headers within the groups
6. If you remove code, check whether that makes some includes obsolete, and remove them if so.

Maybe 6 can be generalized with something like:
"Always check the included files are only necessary for the file in which they are included. Try to avoid excessive including especially in header files."

Reply:
Added.